### PR TITLE
RCHAIN-3184: Use strongly typed keys

### PR DESCRIPTION
**Pull request by JosDenmark**
### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3184

### CODE REVIEW

**Comment by JosDenmark** [comment]
bors try

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RegistrySigGen.scala [hunk 1]
```diff
@@ -109,8 +109,8 @@ object Args {
 
     val id =
       unforgeableNameStr
-        .map(Base16.unsafeDecode(_))
-        .getOrElse(RegistrySigGen.generateUnforgeableNameId(keyPair._1, timestamp))
+        .map(Base16.unsafeDecode)
+        .getOrElse(RegistrySigGen.generateUnforgeableNameId(keyPair._2, timestamp))
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/RegistrySigGen.scala]: What is ._1 and ._2 here?

File: casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala [hunk 2]
```diff
@@ -76,7 +75,7 @@ class MultiParentCasperImpl[F[_]: Sync: Concurrent: Sync: ConnectionsCell: Trans
         dag <- blockDag
         _ <- validatorId match {
               case Some(ValidatorIdentity(publicKey, _, _)) =>
-                val sender = ByteString.copyFrom(publicKey)
+            val sender = ByteString.copyFrom(publicKey.bytes)
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala]: please reformat this commit :)

File: crypto/src/main/scala/coop/rchain/crypto/signatures/Ed25519.scala [hunk 3]
```diff
@@ -10,11 +11,11 @@ object Ed25519 extends SignaturesAlg {
   val name: String = "Ed25519".toLowerCase
 
   //TODO: Make use of strongly typed keys
-  def newKeyPair: (Array[Byte], Array[Byte]) = {
+  def newKeyPair: (PrivateKey, PublicKey) = {
```
Comment (ArturGajowy) [inline crypto/src/main/scala/coop/rchain/crypto/signatures/Ed25519.scala]: Shouldn't other methods of SignaturesAlg take and return the PrivateKey/PublicKey wrappers as well?

Comment (ArturGajowy) [reply crypto/src/main/scala/coop/rchain/crypto/signatures/Ed25519.scala]: I guess I'm fine with sticking to what's done in this PR and scheduling this work for later though.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RegistrySigGen.scala (hunk 1 above)
Comment (ArturGajowy) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RegistrySigGen.scala]: So, are you saying the types won't allow otherwise?

I would be surprised if this change didn't break the RegistrySigGen tool to the extent where it can't produce the right registry url *at all* (vs "you need to provide timestamp, sk, and the uname that's going to be used for registration, after you've fixed timestamp and sk").

TL;DR did you test that you can register a new contract / an old contract under new timestamp + sk + uname, and see the tests pass?

File: crypto/src/main/scala/coop/rchain/crypto/util/Sorting.scala [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package coop.rchain.crypto.util
+
+import coop.rchain.crypto.PublicKey
+import coop.rchain.crypto.codec.Base16
+
+object Sorting {
+
+  implicit val byteArrayOrdering: Ordering[Array[Byte]] =
```
Comment (ArturGajowy) [inline crypto/src/main/scala/coop/rchain/crypto/util/Sorting.scala]: can we make this private, and maybe not even implicit?

File: crypto/src/main/scala/coop/rchain/crypto/util/Sorting.scala [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+    Ordering.by(Base16.encode)
+
+  implicit val publicKeyOrdering: Ordering[PublicKey] =
+    (x: PublicKey, y: PublicKey) => byteArrayOrdering.compare(x.bytes, y.bytes)
```
Comment (ArturGajowy) [inline crypto/src/main/scala/coop/rchain/crypto/util/Sorting.scala]: Since Ordering is a ContravariantFunctor, you can make this better like this:

```scala
  import cats.implicits._ //waaaay up in the imports

  val o: Ordering[PublicKey] = Ordering.by(Base16.encode).contramap(_.bytes)
```

Besides satisfying one's FP kink, we also get to not call `.bytes` twice (which begs for an error!) and have no way of swapping the args involuntarily.

:sunglasses:

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RegistrySigGen.scala (hunk 1 above)
Comment (JosDenmark) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RegistrySigGen.scala]: Yes, it's a type error otherwise. This is the same bug that I mentioned to you earlier today. I haven't tested it, no. I guess there should be a test or two for `RegistrySigGen`. I'll add some.

File: crypto/src/main/scala/coop/rchain/crypto/signatures/Ed25519.scala (hunk 3 above)
Comment (JosDenmark) [reply crypto/src/main/scala/coop/rchain/crypto/signatures/Ed25519.scala]: They do, actually. They're implemented in `SignaturesAlg`, since they're just wrappers.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RegistrySigGen.scala (hunk 1 above)
Comment (JosDenmark) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RegistrySigGen.scala]: @KentShikama `._1` is `PrivateKey`, and `._2` is `PublicKey`.

File: casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala (hunk 2 above)
Comment (JosDenmark) [reply casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala]: 👍

File: crypto/src/main/scala/coop/rchain/crypto/util/Sorting.scala (hunk 4 above)
Comment (JosDenmark) [reply crypto/src/main/scala/coop/rchain/crypto/util/Sorting.scala]: Just because it isn't used anywhere?

File: crypto/src/main/scala/coop/rchain/crypto/util/Sorting.scala (hunk 5 above)
Comment (JosDenmark) [reply crypto/src/main/scala/coop/rchain/crypto/util/Sorting.scala]: 👍

File: crypto/src/main/scala/coop/rchain/crypto/util/Sorting.scala (hunk 4 above)
Comment (JosDenmark) [reply crypto/src/main/scala/coop/rchain/crypto/util/Sorting.scala]: 👍
